File: frameworks/dotnet/carter/UnitTests/CorsTests.cs
namespace UnitTests;

public sealed class CorsTests(CarterApp app) : IClassFixture<CarterApp>
{
    private static readonly string Origin = (string)Expected.Settings["cors"]!["origin"]!;

    // rb:test cors.preflight
    [Fact]
    [Trait("corpus", "cors.preflight")]
    public async Task The_feature_answers_a_preflight_before_any_handler()
    {
        using HttpResponseMessage response = await Preflight(Origin);

        Assert.Equal(HttpStatusCode.NoContent, response.StatusCode);
        Assert.Equal(Origin, Answer.Header(response, "access-control-allow-origin"));
        Assert.Equal("x-rb-tenant", Answer.Header(response, "access-control-allow-headers"));
        Assert.Equal("600", Answer.Header(response, "access-control-max-age"));
        Assert.Null(Answer.Header(response, "x-rb-serial"));
    }

    // rb:test cors.disallowed
    [Fact]
    [Trait("corpus", "cors.disallowed")]
    public async Task A_preflight_from_another_origin_is_not_allowed()
    {
        using HttpResponseMessage response = await Preflight("https://elsewhere.example.net");

        Assert.Null(Answer.Header(response, "access-control-allow-origin"));
    }

    // rb:test cors.request,cors.vary
    [Fact]
    [Trait("corpus", "cors.request")]
    [Trait("corpus", "cors.vary")]
    public async Task The_request_itself_reaches_the_handler_and_varies_on_origin()
    {
        using HttpRequestMessage request = new(HttpMethod.Get, "/cors/small");
        request.Headers.Add("origin", Origin);
        request.Headers.Add("x-rb-tenant", "qwertyuiopas");

        using HttpResponseMessage response = await app.CreateClient().SendAsync(request);

        await Answer.Is(Expected.Json("items.small.json"), response);
        Assert.Equal(Origin, Answer.Header(response, "access-control-allow-origin"));
        Assert.Equal("Origin", Answer.Header(response, "vary"));
        Assert.NotNull(Answer.Header(response, "x-rb-serial"));
    }

    // rb:test cors.scoped
    [Fact]
    [Trait("corpus", "cors.scoped")]
    public async Task A_route_outside_cors_gets_no_policy()
    {
        using HttpRequestMessage request = new(HttpMethod.Get, "/json/small");
        request.Headers.Add("origin", Origin);

        using HttpResponseMessage response = await app.CreateClient().SendAsync(request);

        Assert.Null(Answer.Header(response, "access-control-allow-origin"));
    }

    private Task<HttpResponseMessage> Preflight(string origin)
    {
        HttpRequestMessage request = new(HttpMethod.Options, "/cors/small");
        request.Headers.Add("origin", origin);
        request.Headers.Add("access-control-request-method", "GET");
        request.Headers.Add("access-control-request-headers", "x-rb-tenant");
        return app.CreateClient().SendAsync(request);
    }
}

File: frameworks/java/micronaut/UnitTests/src/test/java/unittests/CorsTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.regex.Pattern;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

class CorsTests extends MicronautApp {

    private static final String ORIGIN = Expected.settings().get("cors").get("origin").getStringValue();

    // rb:test cors.preflight
    @Test
    @Tag("cors.preflight")
    void theFilterAnswersAPreflightBeforeAnyHandler() throws Exception {
        HttpResponse<byte[]> response = preflight(ORIGIN);

        assertEquals(200, response.statusCode());
        assertEquals(ORIGIN, Answer.header(response, "access-control-allow-origin"));
        assertEquals("x-rb-tenant", Answer.header(response, "access-control-allow-headers"));
        assertEquals("600", Answer.header(response, "access-control-max-age"));
        assertNull(Answer.header(response, "x-rb-serial"));
    }

    // rb:test cors.disallowed
    @Test
    @Tag("cors.disallowed")
    void aPreflightFromAnotherOriginIsNotAllowed() throws Exception {
        assertNull(Answer.header(preflight("https://elsewhere.example.net"), "access-control-allow-origin"));
    }

    // rb:test cors.request,cors.vary
    @Test
    @Tag("cors.request")
    @Tag("cors.vary")
    void theRequestItselfReachesTheHandlerAndVariesOnOrigin() throws Exception {
        HttpResponse<byte[]> response = get("/cors/small", "origin", ORIGIN, "x-rb-tenant", "qwertyuiopas");

        Answer.is(Expected.json("items.small.json"), response);
        assertEquals(ORIGIN, Answer.header(response, "access-control-allow-origin"));
        assertTrue(Pattern.compile("(^|,)\\s*origin\\s*(,|$)", Pattern.CASE_INSENSITIVE).matcher(String.join(",", response.headers().allValues("vary"))).find());
        assertNotNull(Answer.header(response, "x-rb-serial"));
    }

    /** The server's CORS configuration has no path of its own, which is why rb.json skips cors.scoped. */
    // rb:test cors.scoped
    @Test
    @Tag("cors.scoped")
    void thePolicyCoversEveryRoute() throws Exception {
        assertEquals(ORIGIN, Answer.header(get("/json/small", "origin", ORIGIN), "access-control-allow-origin"));
    }

    private HttpResponse<byte[]> preflight(String origin) throws Exception {
        return send(request("/cors/small", "origin", origin, "access-control-request-method", "GET", "access-control-request-headers", "x-rb-tenant")
            .method("OPTIONS", HttpRequest.BodyPublishers.noBody()));
    }
}
